src/servers/src/server.rs

```diff
@@ -19,24 +19,76 @@ use std::sync::Arc;
 use async_trait::async_trait;
 use common_runtime::Runtime;
 use common_telemetry::logging::{error, info};
-use futures::future::{AbortHandle, AbortRegistration, Abortable};
+use futures::future::{try_join_all, AbortHandle, AbortRegistration, Abortable};
 use snafu::{ensure, ResultExt};
-use tokio::sync::Mutex;
+use tokio::sync::{Mutex, RwLock};
 use tokio::task::JoinHandle;
 use tokio_stream::wrappers::TcpListenerStream;
 
 use crate::error::{self, Result};
 
 pub(crate) type AbortableStream = Abortable<TcpListenerStream>;
 
-pub type ServerHandlers = HashMap<String, ServerHandler>;
-
 pub type ServerHandler = (Box<dyn Server>, SocketAddr);
 
-pub async fn start_server(server_handler: &ServerHandler) -> Result<Option<SocketAddr>> {
-    let (server, addr) = server_handler;
-    info!("Starting {} at {}", server.name(), addr);
-    server.start(*addr).await.map(Some)
+/// [ServerHandlers] is used to manage the lifecycle of all the services like http or grpc in the GreptimeDB server.
+#[derive(Clone, Default)]
+pub struct ServerHandlers {
+    handlers: Arc<RwLock<HashMap<String, ServerHandler>>>,
+}
+
+impl ServerHandlers {
+    pub fn new() -> Self {
+        Self {
+            handlers: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub async fn insert(&self, handler: ServerHandler) {
+        let mut handlers = self.handlers.write().await;
+        handlers.insert(handler.0.name().to_string(), handler);
+    }
+
+    /// Finds the __actual__ bound address of the service by its name.
+    ///
+    /// This is useful in testing. We can configure the service to bind to port 0 first, then start
+    /// the server to get the real bound port number. This way we avoid doing careful assignment of
+    /// the port number to the service in the test.
+    ///
+    /// Note that the address is guaranteed to be correct only after the `start_all` method is
+    /// successfully invoked. Otherwise you may find the address to be what you configured before.
+    pub async fn addr(&self, name: &str) -> Option<SocketAddr> {
+        let handlers = self.handlers.read().await;
+        handlers.get(name).map(|x| x.1)
+    }
+
+    /// Starts all the managed services. It will block until all the services are started. And it will
+    /// set the actual bound address to the service.
```

Comment:
nit:
```suggestion
    /// Starts all the managed services. It will block until all the services are started. 
    /// And it will set the actual bound address to the service.
```